File: code/src/main/java/chapter1/java/Session.java
package chapter1.java;

import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import static java.util.Arrays.asList;


public class Session {

    public final String title;
    @Nullable // <1>
    public final String subtitleOrNull;
    public final List<Presenter> presenters; // <2>

    public Session(String title, @Nullable String subtitle, List<Presenter> presenters) {
        this.title = title;
        this.subtitleOrNull = subtitle;
        this.presenters = Collections.unmodifiableList(new ArrayList<>(presenters)); // <3>
    }

    public String getTitle() {
        return title;
    }

    @Nullable
    public String getSubtitleOrNull() {
        return subtitleOrNull;
    }

    public List<Presenter> getPresenters() {
        return presenters;
    }

    // <4>
    public Session withPresenters(List<Presenter> newLineUp) {
        return new Session(title, subtitleOrNull, newLineUp);
    }

    public Session withTitle(String newTitle) {
        return new Session(newTitle, subtitleOrNull, presenters);
    }

    public Session withSubtitle(@Nullable String newSubtitle) {
        return new Session(title, newSubtitle, presenters);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Session session = (Session) o;
        return Objects.equals(title, session.title) &&
                Objects.equals(subtitleOrNull, session.subtitleOrNull) &&
                Objects.equals(presenters, session.presenters);
    }

    @Override
    public int hashCode() {
        return Objects.hash(title, subtitleOrNull, presenters);
    }

    @Override
    public String toString() {
        return "chapter1.java.Session{" +
                "title='" + title + '\'' +
                ", subtitle='" + subtitleOrNull + '\'' +
                ", presenters=" + presenters +
                '}';
    }
}

File: code/src/main/kotlin/chapter1/kotlin/Session.kt
package chapter1.kotlin

import chapter1.java.Presenter
import java.util.*

class Session(
    val title: String,
    val subtitleOrNull: String?,
    presenters: List<Presenter>
) {

    val presenters : List<Presenter>

    constructor(title: String, subtitle: String?, vararg presenters: Presenter) : this(
        title,
        subtitle,
        presenters.asList()
    )

    fun withPresenters(newLineUp: List<Presenter>): Session {
        return Session(title, subtitleOrNull, newLineUp)
    }

    fun withTitle(newTitle: String): Session {
        return Session(newTitle, subtitleOrNull, presenters)
    }

    fun withSubtitle(newSubtitle: String?): Session {
        return Session(title, newSubtitle, presenters)
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val session = o as Session
        return title == session.title &&
            subtitleOrNull == session.subtitleOrNull &&
            presenters == session.presenters
    }

    override fun hashCode(): Int {
        return Objects.hash(title, subtitleOrNull, presenters)
    }

    override fun toString(): String {
        return "chapter1.java.Session{" +
            "title='" + title + '\'' +
            ", subtitle='" + subtitleOrNull + '\'' +
            ", presenters=" + presenters +
            '}'
    }

    init {
        this.presenters = Collections.unmodifiableList(ArrayList(presenters))
    }
}